### ISSUE TITLE
HDF5 Support

### ISSUE THREAD

**Issue by Young-Todd** [scikit-optimize/scikit-optimize#401]
Hi all,

If possible, I would like to add support for saving optimization results in HDF5 format. I have been running many hyperparameter optimization procedures and have been thinking that, rather dumping many pickle files, it would be nice to write to HDF5 with h5py. 

Has anyone had a go at this? Would there be any immediate pitfalls/concerns for this approach?

**Comment by betatim** [comment]
I don't think I know anyone who has tried/worked on this before.

Immediate thoughts: can we get some experience from the `joblib` people about hdf5 vs joblib for storing large scikit-learn models. I think I read in their docs that they thought about this and optimised it a bit.

The other question would be: can we integrate it with the current `utils.dump/load` and avoid adding hdf5 as a dependency (if reducing number of dependencies is something we want to do).

**Comment by Young-Todd** [comment]
I do not have much experience with `joblib`, and would be interested if anyone would have some input when it comes to storing the results of the hyperparameter optimization (including the objective function). 

Trying a naive approach, just dumping the results of the Gaussian process example on https://scikit-optimize.github.io/notebooks/hyperparameter-optimization.html with 

```python
from sklearn.externals import joblib
import numpy as np
from skopt import gp_minimize

# ... 
res_gp = gp_minimize(objective, space, n_calls=150, random_state=0, verbose=True)

joblib.dump(res_gp, "joblib_test.pkl")
returned_results = joblib.load('joblib_test.pkl')
```

Which returns with the following error, where there is no attribute 'objective' to load

![joblib_objective_error](https://user-images.githubusercontent.com/16848331/27189417-4663526e-51bf-11e7-8018-a4f51ef79507.png)

I know the documentation covers the issue of storing the optimization objective and that can be alleviated by setting the keyword argument `store_objective=False`. If possible, I would like to come up with a way that would allow that objective to be saved and make it work with the existing `utils.dump/load`.

**Comment by Young-Todd** [comment]
Are there any main features from `skopt` that will be lost without having saved the objective function, say any of the plotting, etc.?

**Comment by betatim** [comment]
https://scikit-optimize.github.io/#skopt.dump and https://scikit-optimize.github.io/#skopt.load should work with objectives that can be pickled. The reason for having `store_objective=False` is to support cases where the objective is something that can not be pickled.

Can't think of anything off the top of my head that wouldn't work. I think for cases where the objective can't be pickled people know how to "restore" the objective function through some other means.

**Comment by Young-Todd** [comment]
So if the `skopt.dump` does not throw an exception, then `skopt.load` should work fine? If that is the case, I am not sure why I got the error above. It only ran with an error when loading it back from disk and did not have an issue when dumping it.

**Comment by betatim** [comment]
Yes, if `skopt.load` raises an exception I would consider that a bug.

In the screenshot above you used `joblib` directly. Can you try again via `skopt`?

**Comment by Young-Todd** [comment]
After using `skopt.dump` and `skopt.dump`, I am running into `AttributeError: module '__main__' has no attribute 'objective'`. I am testing it on the Gaussian process example (without making any changes) here: https://scikit-optimize.github.io/notebooks/hyperparameter-optimization.html

![load_attribute_error](https://user-images.githubusercontent.com/16848331/27291130-6594476c-54dd-11e7-9696-c07fd17779a3.png)

**Comment by MechCoder** [comment]
Your traceback suggests it has something to do with Python3. Upon some googling, it should work if you do "from script import objective" before unpickling the object.

The reason is given here (https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled)

```
Note that functions (built-in and user-defined) are pickled by “fully qualified” name reference
not by value. [2].
This means that only the function name is pickled, along with the name of the module
the function is defined in.
Neither the function’s code, nor any of its function attributes are pickled. Thus the defining module must be importable in the unpickling environment
and the module must contain the named object, otherwise an exception will be raised. [3]
```

So in a fresh environment, it cannot find the function objective in `__main__` and hence you need to import it.

**Comment by MechCoder** [comment]
Coming to the original issue, why and what do you want to store in the hdf5 format?

If you have a good rationale for that and you are willing to write it, we can definitely support something like reading and writing into hdf5 file formats. I'm not super familiar with it myself.

**Comment by Young-Todd** [comment]
Thanks @MechCoder, after importing my objective function everything works well. 

Considering the hdf5 format, maybe I should start with the research project I am working on. I am a student intern at Oak Ridge National Lab and have been working on hyperparameter optimization for my group's various machine learning models (mostly large convolutional nets and recurrent nets) using the lab's supercomputers. My approach has been to break down the search space into small overlapping subregions over which I can run the sequential model based optimization procedures. It is my hope that distributing the process in this way would allow us to find a global optima faster than starting with the larger search space. More interestingly, I would like to see if we can find many reasonable local optima. If we do find many reasonable local optima, I would like compare how those settings of hyperparameters generalize compared with the global optima found.

Depending upon the size of the original search space, my approach can run in the 1000's of parallel optimization procedures. My group has asked me to look into storing the results as hdf5 as a way of organizing these many results. The file format is pretty popular with the HPC groups at the lab. I am not very familiar with hdf5 myself, and I am certainly not married to it. What I do need is a way of organizing all of these results for analysis and to later publish. 

What are your thoughts? If the scikit-optimize team would be interested in this approach, I would be more than happy to share/contribute in any way.

**Comment by MechCoder** [comment]
Sure, good to hear that! feel free to share your code once you have a working version that dumps the results into a hdf5 format in the form of a pull request.

**Comment by betatim** [comment]
hdf5 seems pretty popular and well tested. With some technical finesse we can organise for `dump`/`load` to support joblib and hdf5 without making hdf5 a new dependency.

Another cool addition (for your problem) would be a callback like https://keras.io/callbacks/#modelcheckpoint built on top of `dump`/`load`
